src/UNIV2LPOracle.sol

```diff
@@ -264,14 +294,68 @@ contract UNIV2LPOracle {
         );
     }
 
-    function poke() external stoppable {
-        require(pass(), "UNIV2LPOracle/not-passed");
-        uint128 val = seek();
-        require(val != 0, "UNIV2LPOracle/invalid-price");
-        cur = nxt;
-        nxt = Feed(val, 1);
-        zzz = uint64(block.timestamp);
-        emit Value(cur.val, nxt.val);
+    function poke() external {
+
+        // Ensure a single SLOAD while avoiding solc's excessive bitmasking bureaucracy.
+        uint256 _zph;
+        uint256 _hop;
+        {
+            uint256 _stopped;  // block-scoping _stopped here saves a little gas
+            assembly {
+                let _slot1 := sload(1)
+                _stopped   := and(_slot1,         0xff    )
+                _hop       := and(shr(8, _slot1), 0xffffff)
+                _zph       := shr(32, _slot1)
+            }
+
+            // When stopped, values are set to zero and should remain such; thus, disallow updating in that case.
+            require(_stopped == 0, "UNIV2LPOracle/is-stopped");
+        }
+
+        // Equivalent to requiring that pass() returns true.
+        // The logic is repeated instead of calling pass() to save gas
+        // (both by eliminating an internal call here, and allowing pass to be external).
+        require(block.timestamp >= _zph, "UNIV2LPOracle/not-passed");
+
+        uint128 _val = seek();
+        require(_val != 0, "UNIV2LPOracle/invalid-price");
+        Feed memory _cur = nxt;  // This memory value is used to save an SLOAD later.
+        cur = _cur;
+        nxt = Feed(_val, 1);
+
+        // The below is equivalent to:
+        //
+        //    zph = block.timestamp + hop
+        //
+        // but ensures no extra SLOADs are performed.
+        //
+        // Even if _hop = (2^24 - 1), the maximum possible value, add(timestamp(), _hop)
+        // will not overflow (even a 224 bit value) for a very long time.
+        //
+        // Also, we know stopped was zero, so there is no need to account for it explicitly here.
+        assembly {
+            sstore(
+                1,
+                add(
+                    shl(                          // zph value starts 32 bits in
+                        32,
+                        add(timestamp(), _hop)
+                    ),
+                    shl(                          // hop value starts 8 bits in
+                        8,
+                        _hop
+                    )
+                )
+            )
+        }
+
+        // Equivalent to emitting Value(cur.val, nxt.val), but averts two extra SLOADs.
+        emit Value(_cur.val, _val);
```

Comment:
how about renaming `_val` to `_nxt`?

Reply:
I think that's misleading, because it makes one think it's equal to `nxt`, which is a packed structure consisting of a `val` and a `has`.